### DISCUSSION TITLE
jax.grad yields nan of jnp.sqrt(array)?

### DISCUSSION THREAD

**Discussion by TomClarkMassSpec** [jax-ml/jax#17263] [Q&A]
working with 64bit and isolated a problem down to one line which I abbreviate here:

```
value_inside_exp = multiplier * jnp.sqrt(arr)
result = jnp.exp(value_inside_exp)
```
example input:
```
multiplier
Out[93]: DeviceArray(-3.55448731e-13, dtype=float64)

Out[94]: 
 jnp.sqrt(arr)
DeviceArray([0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             1.86344989, 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
...
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.02569974, 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.36218982,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ,
             0.        , 0.        , 0.        , 0.        , 0.        ],            dtype=float64)

```

So if there was problem with the value then I could use jnp.where or otherwise to retain a value at extremes. Assuming this is about extreme values, the grad returns nan. I've tried clipping the 'value_inside_exp' but that just affects the value while the grad still returns nan values in the result. Is there a way to control the grad output (ie set to zero when values or grad are really small)?

I do think nan are showing up in the gradient array due to scale of numbers as the input 'arr' above is operated on elsewhere without issue, while the variables in grad wrt variables array that deliver nan are the variables used in building this padded array.... unless I'm missing another reason for the nan's. 

Every function is @jax.jit.

**Comment by jakevdp** [comment]
I don't quite understand the question – here's the code I constructed based on what you provided in your question:
```python
import jax
import jax.numpy as jnp

jax.config.update('jax_enable_x64', True)

multiplier = jnp.float64(-3.55448731e-13)
arr = jnp.array([0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 1.86344989, 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.02569974, 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.36218982,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ,
                 0.        , 0.        , 0.        , 0.        , 0.        ])

value_inside_exp = multiplier * jnp.sqrt(arr)
result = jnp.exp(value_inside_exp)
print(result)
# [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
#  1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
#  1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
#  1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
#  1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
```
This seems like reasonable output – what should I be looking for?

**Reply by TomClarkMassSpec** [reply]
Hi Jake,

Yes, this data does produce that output for those inputs. I'm struggling to avoid the nans in grad when I apply jax.value_and_grad (albeit to a much larger function). I think I'm closer as the problem appears to be with jax.numpy.sqrt(array) or my use of it:

```
part_term_2 = jnp.sqrt(array_of_micelles) yields,
Gradient: [DeviceArray(nan, dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(nan, dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64)]
```

but if I replace with:

```
part_term_2 = (array_of_micelles), this yields a sensible answer:
Gradient: [DeviceArray(-38.24011697, dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(10.31770934, dtype=float64), DeviceArray(8.20919978, dtype=float64), DeviceArray(34.28429711, dtype=float64), DeviceArray(17.51066209, dtype=float64), DeviceArray(13.67384336, dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(14.58782719, dtype=float64), DeviceArray(16.53114765, dtype=float64), DeviceArray(15.4385741, dtype=float64), DeviceArray(11.04987305, dtype=float64), DeviceArray(15.16785687, dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64), DeviceArray(0., dtype=float64)]  

```

The latter being correct and devoid of nan's. Sorry for the lack of clarity -> my problem is with jax.grad yielding nans. I am using jax.nump.sqrt to take the square root element wise of array elements which yields the correct value but not the correct grad result.

Thanks,
Tom

**Reply by jakevdp** [reply]
What code did you run to get the output after `Gradient:`?

(I could make a guess, but in general I've found that guessing what code people ran for bug reports or questions does not lead to helpful discussion)

**Reply by TomClarkMassSpec** [reply]
I slashed and burned everything after, only returning the variables of interest:
```
@jax.jit
def protein_data_transformations_for_terminal_speed(custom_tree):
    # Unpack relevant pytree variables and constants

    ....   
    
    #micelle_speed_term2 = jnp.exp(-C5 * radius_micelle *... * jnp.sqrt(array_of_micelles))
    part_term_2 = (array_of_micelles)
    return part_term_2

def distance_prediction(custom_tree):
        # terminal speed of protein-SDS complex with ot without gel fiber interaction
    micelles_speed_term_2 = protein_data_transformations_for_terminal_speed(custom_tree) 
    distance_prediction = jnp.sum(micelles_speed_term_2)
    return distance_prediction

def calc_individual_y(variables_list, constants_tuple, padded_seq, protein_length):
    custom_tree = CustomPyTree(
        variables_list,
        constants_tuple,
        (
            padded_seq,
            protein_length
        )
    )
    y = distance_prediction(custom_tree)
    return y

def wrapped_y_function(variables, constants_tuple, padded_seq, protein_length):
    v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11, v12, v13, v14, v15, v16, v17, v18, v19, v20, v21, v22, v23, v24, v25, v26, v27, v28, v29 = variables
    return calc_individual_y([v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11, v12, v13, v14, v15, v16, v17, v18, v19, v20, v21, v22, v23, v24, v25, v26, v27, v28, v29], constants_tuple, padded_seq, protein_length)

# Value and gradient function
value_and_grad_fn = jax.value_and_grad(wrapped_y_function, argnums=0)

# Evaluate value and gradient at some point

#value, gradient = value_and_grad_fn(variables_at_some_point, constants_at_some_point, data_array_at_some_point)
value, gradient = value_and_grad_fn(variables_at_some_point, constants_tuple, padded_seq_at_some_point, protein_length_at_some_point)

print("Value:", value)
print("Gradient:", gradient)
```

**Reply by jakevdp** [reply]
There are several undefined variables in your code: `array_of_micelles`, `CustomPyTree`, `variables_at_some_point`, `constants_tuple`, `padded_seq_at_some_point`, `protein_length_at_some_point`... etc.

I'd love to help, but I can't help if I don't know what code you're running!

**Reply by TomClarkMassSpec** [reply]
I don't think I can drop a thousand lines of code. But  I figured it out and will leave the answer. In short I have non-functional zeros such as due to padding to maintain array shape. Sqrt of 0. is 0. and gradient of that is inf. Further down the code, the inf. yields nan for jax.grad. The fix:
```
    masked_sds_cluster_quant = sdscluster_quant>0
    multiplier = -C5 * radius_micelle *...
    exponential_term_2 = jnp.where(masked_sds_cluster_quant, multiplier*jnp.sqrt(sdscluster_quant),0.)
    micelle_speed_term2 = jnp.exp(exponential_term_2)

```
Thanks for looking

**Reply by jakevdp** [reply]
That makes sense - glad you figured out the issue!

**Reply by keithlinndowning** [reply]
Thanks Tom.  I should have searched for this 4 hours ago, but it took me that long to localize my error to jnp.sqrt.
